webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/microsoft/MicrosoftMTEngine.java

```diff
@@ -0,0 +1,190 @@
+
+        HttpEntity<List<Map<String, String>>> requestEntity = new HttpEntity<>(body, headers);
+
+        try {
+            MicrosoftTextTranslationDTO[] response =
+                    restTemplate.postForObject(
+                            finalUri, requestEntity, MicrosoftTextTranslationDTO[].class);
+
+            return getTranslationsResponseDTO(textSources, response, targetLocaleTags);
+        } catch (HttpClientErrorException e) {
+            // Log only the request body, as the headers contain authentication information.
+            String errorMessage =
+                    String.format(
+                            "Could not Machine Translate posted to: %s with %s: %s and request body: %s",
+                            finalUri,
+                            X_CLIENT_TRACE_ID,
+                            requestEntity.getHeaders().get(X_CLIENT_TRACE_ID),
+                            body);
+            logger.error(errorMessage, e);
+            throw e;
+        }
+    }
+
+    private TranslationsResponse getTranslationsResponseDTO(
+            List<String> textSources, MicrosoftTextTranslationDTO[] response, List<String> targetLocaleTags) {
+        List<TextUnitTranslationGroup> textUnitTranslationGroups = new ArrayList<>();
+
+        // Translation objects in the output for each source string are returned in the same order as provided by the client.
+        for (int i = 0; i < textSources.size(); i++) {
+            String sourceText = textSources.get(i);
+            MicrosoftTextTranslationDTO microsoftTextTranslationDTO = response[i];
+
+            TextUnitTranslationGroup textUnitTranslationGroupDTO = new TextUnitTranslationGroup();
+            textUnitTranslationGroupDTO.setSourceText(sourceText);
+            textUnitTranslationGroupDTO.setTranslations(getTranslationDTOS(targetLocaleTags, microsoftTextTranslationDTO));
+            textUnitTranslationGroups.add(textUnitTranslationGroupDTO);
+        }
+
+        return new TranslationsResponse(textUnitTranslationGroups);
+    }
+
+    private List<Translation> getTranslationDTOS(List<String> targetLocaleTags, MicrosoftTextTranslationDTO microsoftTextTranslationDTO) {
+        List<Translation> translationDTOS = new ArrayList<>();
+        // Within a set of translations for a specific source string, the languages are ordered in the same way as
+        // the order of target languages provided to the API.
+        for (int j = 0; j < targetLocaleTags.size(); j++) {
+            MicrosoftLanguageTranslationDTO microsoftLanguageTranslationDTO = microsoftTextTranslationDTO.getTranslations().get(j);
+            String requestTargetLocale = targetLocaleTags.get(j);
+
+            Translation translationDTO = new Translation();
+            translationDTO.setTranslationSource(getSource());
+            translationDTO.setLocaleTag(requestTargetLocale);
+            translationDTO.setText(microsoftLanguageTranslationDTO.getText());
```

Comment:
nit: I see potential stream everywhere now 😭

Reply:
We can't use streams here (as with some of the other scenarios) as we need to iterate through two data sources at the same time.
The API response from the MT engine gives us back results in a specific order but with their own language tags - and we want to reply back to our users with the original language tags that they gave us for consistency.